Pulsar: add flag to enable transactions and set configuration

nicoloboschi opened the pull request:
* Add `withTransactions()` method to enable transactions on Pulsar container
* Set default version to latest released (2.10.0)
* New docker command that enables the user to easily change configuration parameters
* Added new tests to cover new methods
* Improved the documentation with simple usage and new methods

nicoloboschi commented on modules/pulsar/src/main/java/org/testcontainers/containers/PulsarContainer.java:
```diff
@@ -41,36 +51,65 @@ public PulsarContainer(String pulsarVersion) {
+                    .withStrategy(Wait.forLogMessage(".*Function worker service started.*", 1))
+            );
+        } else if (transactions) {
+            waitingFor(Wait.forHttp(TRANSACTION_TOPIC_ENDPOINT).forStatusCode(200).forPort(BROKER_HTTP_PORT));
```
This is needed because the Pulsar client will try to start the TransactionCoordinatorClient and if the topic doesn't exist yet it will fail. 
So if we remove that this code won't work:

```
try (PulsarContainer pulsar = new PulsarContainer(PULSAR_IMAGE).withTransactions();) {
  try (PulsarClient client = PulsarClient.builder().serviceUrl(pulsarBrokerUrl).enableTransaction(true).build();) {
   ....       
  }
}
```

and you'll get 
```
Caused by: org.apache.pulsar.client.api.transaction.TransactionCoordinatorClientException: org.apache.pulsar.client.api.PulsarClientException$BrokerMetadataException: {"errorMsg":"Namespace not found","reqId":3642126209754600375, "remote":"localhost/127.0.0.1:54629", "local":"/127.0.0.1:54823"}
	at app//org.apache.pulsar.client.api.transaction.TransactionCoordinatorClientException.unwrap(TransactionCoordinatorClientException.java:131)
	at app//org.apache.pulsar.client.api.transaction.TransactionCoordinatorClientException.unwrap(TransactionCoordinatorClientException.java:129)
	at app//org.apache.pulsar.client.impl.transaction.TransactionCoordinatorClientImpl.start(TransactionCoordinatorClientImpl.java:70)
	at app//org.apache.pulsar.client.impl.PulsarClientImpl.<init>(PulsarClientImpl.java:204)
	... 47 more
```

nicoloboschi replied:
btw I created this to improve the Pulsar client https://github.com/apache/pulsar/issues/15996

nicoloboschi commented:
@eddumelendez @kiview PTAL

kiview commented on modules/pulsar/src/main/java/org/testcontainers/containers/PulsarContainer.java:
```diff
@@ -41,36 +51,65 @@ public PulsarContainer(String pulsarVersion) {
+            withConfiguration("transactionCoordinatorEnabled", "true");
+        }
+
+        if (functionsWorkerEnabled) {
```
Is this hierarchy for defining the wait strategy correct? So if `functionsWorkerEnabled` and `transactionsEnabled`, it is enough to wait for `Wait.forLogMessage(".*Function worker service started.*", 1)`? Or would we need a `WaitAllStrategy` that combines the waiting aspects?

eddumelendez replied:
also, setting the defaultStrategy, sounds redundant to me.

nicoloboschi replied:
`WaitAllStrategy` seems the best option, thanks.

@eddumelendez can you explain better? what are you suggesting?

eddumelendez replied:
default strategy is looking is just waiting for the default port configured and later other strategies have been added which already cover the listening port plus the readiness of a specific path. I would like to understand why the defaultStrategy was added

kiview commented on modules/pulsar/src/main/java/org/testcontainers/containers/PulsarContainer.java:
```diff
@@ -15,13 +17,21 @@ public class PulsarContainer extends GenericContainer<PulsarContainer> {
 
     private boolean functionsWorkerEnabled = false;
 
+    private boolean transactions = false;
```
```suggestion
    private boolean transactionsEnabled = false;
```

Just so it is symmetrical to `functionsWorkerEnabled`.

kiview commented on modules/pulsar/src/main/java/org/testcontainers/containers/PulsarContainer.java:
```diff
@@ -15,13 +17,21 @@ public class PulsarContainer extends GenericContainer<PulsarContainer> {
+    /**
+     * See <a href="https://github.com/apache/pulsar/blob/master/pulsar-common/src/main/java/org/apache/pulsar/common/naming/SystemTopicNames.java">SystemTopicNames</a>.
+     */
+    public static final String TRANSACTION_TOPIC_ENDPOINT =
```
Does this need to be `public`?

kiview reviewed (commented):
Thanks for the PR and special thanks for adding the docs @nicoloboschi. I added some general comments and questions 🙂

kiview commented:
Pulsar tests failing on CI:
https://github.com/testcontainers/testcontainers-java/runs/6901742299?check_suite_focus=true#step:7:104

nicoloboschi commented on modules/pulsar/src/main/java/org/testcontainers/containers/PulsarContainer.java:
```diff
@@ -41,36 +51,65 @@ public PulsarContainer(String pulsarVersion) {
+        return this;
+    }
+
+    public PulsarContainer withConfiguration(String name, String value) {
+        return withEnv("PULSAR_PREFIX_" + name, value);
+    }
+
+    public PulsarContainer withConfiguration(Map<String, String> configuration) {
+        configuration.forEach((name, value) -> withEnv("PULSAR_PREFIX_" + name, value));
+        return this;
+    }
```
it is helpful because the right way to add a configuration property to Pulsar standalone is to set env with PULSAR_PREFIX_<configname>

otherwise the users have to remember to add the prefix

nicoloboschi replied:
Unfortunately I haven't found anything on the docs. 
This is the issue that implemented it: https://github.com/apache/pulsar/pull/3858

This is the script that does the work: https://github.com/apache/pulsar/blob/master/docker/pulsar/scripts/apply-config-from-env-with-prefix.py
and indeed I had to change the docker cmd in this pull: https://github.com/testcontainers/testcontainers-java/pull/5479/files#diff-87d7d34dfb8435aa9ab0532510e5109ae787d9b8126e398300293bf705b3cf4fR92

this is usually how you deploy real Pulsar cluster but the Pulsar standalone is meant for testing only so you need to explicity call the script

nicoloboschi replied:
upcoming pull that explain to PULSAR_PREFIX_ usage: https://github.com/apache/pulsar/pull/16085/files

kiview replied:
Thanks for the explanation. So I wasn't too incompetent to find the docs on `PULSAR_PREFIX`, they were simply not existing at this point in time 😅 

If I am honest, once the docs are in place in https://github.com/apache/pulsar/pull/16085, I think having the users configure it explicitly simply via `withEnv` seems Testcontainers idiomatic to me.

nicoloboschi replied:
What if I rename it `withPulsarEnv()`? 
I'm pretty sure people will forget to use the prefix.

Also, a nuance of Pulsar, if you set it without the prefix some properties work and other not 😞

kiview replied:
I can imagine that users might forget the prefix, but I can also imagine that users make typos if those values are just configured as Strings. Some other Testcontainers modules tend to use `enums` to mitigate this, but this approach comes with its own set of maintenance overhead.

In general, configuring Docker containers via ENV is idiomatic for Docker and for Testcontainers alike, and by removing the method, we don't unnecessarily increase the API size of the module. It is also pretty normal to have the technology-specific settings prefixed in some way (but it must admit the combination of `PULSAR_PREFIX` + camelCase config is a bit unusual to me).

Therefore, I think configuring those settings using the existing `withEnv()` method is the least surprising for Testcontainers users.

nicoloboschi commented:
@kiview thanks for reviewing. I fixed your comments, could you review it again?

eddumelendez commented on modules/pulsar/src/main/java/org/testcontainers/containers/PulsarContainer.java:
```diff
@@ -41,36 +53,70 @@ public PulsarContainer(String pulsarVersion) {
+
+        withCommand("/bin/bash", "-c", standaloneBaseCommand);
+
+        List<WaitStrategy> waitStrategyList = new ArrayList<>();
+        waitStrategyList.add(this.waitStrategy);
+        waitStrategyList.add(
+            Wait.forHttp(METRICS_ENDPOINT).forStatusCode(200).forPort(BROKER_HTTP_PORT)
+        );
+        if (transactionsEnabled) {
+            withConfiguration("transactionCoordinatorEnabled", "true");
+            waitStrategyList.add(
+                Wait.forHttp(TRANSACTION_TOPIC_ENDPOINT).forStatusCode(200).forPort(BROKER_HTTP_PORT)
+            );
+        }
+        if (functionsWorkerEnabled) {
+            waitStrategyList.add(
+                Wait.forLogMessage(".*Function worker service started.*", 1)
+            );
+        }
+        final WaitAllStrategy compoundedWaitStrategy = new WaitAllStrategy();
+        waitStrategyList.forEach(compoundedWaitStrategy::withStrategy);
```
```suggestion
        List<WaitStrategy> waitStrategies = new ArrayList<>();
        waitStrategies.add(this.waitStrategy);
        waitStrategies.add(
            Wait.forHttp(METRICS_ENDPOINT).forStatusCode(200).forPort(BROKER_HTTP_PORT)
        );
        if (transactionsEnabled) {
            withConfiguration("transactionCoordinatorEnabled", "true");
            waitStrategies.add(
                Wait.forHttp(TRANSACTION_TOPIC_ENDPOINT).forStatusCode(200).forPort(BROKER_HTTP_PORT)
            );
        }
        if (functionsWorkerEnabled) {
            waitStrategies.add(
                Wait.forLogMessage(".*Function worker service started.*", 1)
            );
        }
        final WaitAllStrategy compoundedWaitStrategy = new WaitAllStrategy();
        waitStrategies.forEach(compoundedWaitStrategy::withStrategy);
```

eddumelendez commented on modules/pulsar/src/test/java/org/testcontainers/containers/PulsarContainerTest.java:
```diff
@@ -19,36 +23,102 @@ public class PulsarContainerTest {
+            assertThat(pulsar.getEnvMap().get("PULSAR_PREFIX_clusterName")).isEqualTo("mycluster");
+
+            pulsar.withConfiguration(Maps.newHashMap("clusterName", "mycluster2", "maxTopicsPerNamespace", "10"));
+            assertThat(pulsar.getEnvMap().get("PULSAR_PREFIX_clusterName")).isEqualTo("mycluster2");
+            assertThat(pulsar.getEnvMap().get("PULSAR_PREFIX_maxTopicsPerNamespace")).isEqualTo("10");
```
```suggestion
            assertThat(pulsar.getEnvMap()).containsEntry("PULSAR_PREFIX_clusterName", "mycluster2");
            assertThat(pulsar.getEnvMap()).containsEntry("PULSAR_PREFIX_maxTopicsPerNamespace", "10");
```

eddumelendez commented on modules/pulsar/src/test/java/org/testcontainers/containers/PulsarContainerTest.java:
```diff
@@ -19,36 +23,102 @@ public class PulsarContainerTest {
+    public void testConfiguration() throws Exception {
+        try (PulsarContainer pulsar = new PulsarContainer(PULSAR_IMAGE)) {
+            pulsar.withConfiguration("clusterName", "mycluster");
+            assertThat(pulsar.getEnvMap().get("PULSAR_PREFIX_clusterName")).isEqualTo("mycluster");
```
```suggestion
            assertThat(pulsar.getEnvMap()). containsEntry("PULSAR_PREFIX_clusterName", "mycluster");
```

eddumelendez commented on modules/pulsar/src/test/java/org/testcontainers/containers/PulsarContainerTest.java:
```diff
@@ -19,36 +23,102 @@ public class PulsarContainerTest {
-
-            assertThatThrownBy(() -> pulsarAdmin.functions().getFunctions("public", "default"))
-                .isInstanceOf(PulsarAdminException.class);
+            try (PulsarAdmin pulsarAdmin = PulsarAdmin.builder().serviceHttpUrl(pulsar.getHttpServiceUrl()).build();) {
```
```suggestion
            try (PulsarAdmin pulsarAdmin = PulsarAdmin.builder().serviceHttpUrl(pulsar.getHttpServiceUrl()).build()) {
```

eddumelendez commented on modules/pulsar/src/test/java/org/testcontainers/containers/PulsarContainerTest.java:
```diff
@@ -19,36 +23,102 @@ public class PulsarContainerTest {
             pulsar.start();
 
-            PulsarAdmin pulsarAdmin = PulsarAdmin.builder().serviceHttpUrl(pulsar.getHttpServiceUrl()).build();
+            try (PulsarAdmin pulsarAdmin = PulsarAdmin.builder().serviceHttpUrl(pulsar.getHttpServiceUrl()).build();) {
```
```suggestion
            try (PulsarAdmin pulsarAdmin = PulsarAdmin.builder().serviceHttpUrl(pulsar.getHttpServiceUrl()).build()) {
```

eddumelendez commented on modules/pulsar/src/test/java/org/testcontainers/containers/PulsarContainerTest.java:
```diff
@@ -19,36 +23,102 @@ public class PulsarContainerTest {
+            pulsar.start();
 
-            assertThat(pulsarAdmin.functions().getFunctions("public", "default")).hasSize(0);
+            try (PulsarAdmin pulsarAdmin = PulsarAdmin.builder().serviceHttpUrl(pulsar.getHttpServiceUrl()).build();) {
```
```suggestion
            try (PulsarAdmin pulsarAdmin = PulsarAdmin.builder().serviceHttpUrl(pulsar.getHttpServiceUrl()).build()) {
```

eddumelendez reviewed (commented):
thanks @nicoloboschi ! Really glad to see PulsarContainer will make people life easier :) 

I have added minor suggestions and opened the topic about the strategies order.

kiview commented:
CI is failing because of Spotless violations, you can run `./gradlew :pulsar:spotlessApply' to fix them.

kiview reviewed (commented):
LGTM besides the open discussion about `withConfiguration`. 

I would prefer if we switch it to `withEnv`, add an usage example to our docs, and further link the Pulsar docs once https://github.com/apache/pulsar/pull/16085 is live.

nicoloboschi commented:
@kiview I think you know better the testcontainers users than me
I removed the custom configuration method, added the constant inside the PulsarContainer with javadoc, I added a paragraph in the website that points to the Pulsar website for the Docker setup. 
Also the explanation about PULSAR_PREFIX_ is already live: https://pulsar.apache.org/docs/getting-started-docker/#start-pulsar-in-docker

eddumelendez commented on modules/pulsar/src/main/java/org/testcontainers/containers/PulsarContainer.java:
```diff
@@ -15,13 +19,26 @@ public class PulsarContainer extends GenericContainer<PulsarContainer> {
-    private static final String DEFAULT_TAG = "2.2.0";
+    private static final String DEFAULT_TAG = "2.10.0";
+
+    /**
+     * Environment variable prefix needed to configure Pulsar standalone. See <a href="https://pulsar.apache.org/docs/next/getting-started-docker/">the guide</a>.
+     */
+    public static final String ENV_PULSAR_PREFIX = "PULSAR_PREFIX_";
```
I think pulsar users will know about this and will be adding the prefix by themselves instead of relying on this constant.. IMHO, it should be removed.

nicoloboschi commented:
@kiview @eddumelendez PTAL again, I hope it's the latest round 🤞

kiview reviewed (approved):
LGTM, thanks for the great collaboration @nicoloboschi 🙌
